lib/spack/docs/packaging_guide.rst

```diff
@@ -4054,36 +4054,227 @@ File functions
+
+   @classmethod
+   def determine_version(cls, exe):
+       """Return either the version of the executable passed as argument
+       or ``None`` if the version cannot be determined
+       """
+
+This method receives as input the path to a single executable and must return
+as output its version as a string, or ``None`` if the version cannot be determined
+or the executable is not related to the package.
+Implementing the three steps above is mandatory, and gives the package the
+basic ability to detect if a spec is present on the system at a given version.
+
+.. note::
+   Any executable for which the ``determine_version`` method returns ``None``
+   will be discarded and won't appear in later stages of the workflow described below.
+
+^^^^^^^^^^^^^^^^^^^^^^^^
+Additional functionality
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Besides the three mandatory steps described above, there are also optional
+methods that can be implemented to either increase the amount of details
+being detected or improve the robustness of the detection logic in a package.
+
+""""""""""""""""""""""""""""""
+Variants and custom attributes
+""""""""""""""""""""""""""""""
+
+The ``determine_variants`` method can be optionally implemented in a package
+to detect additional details of the spec:
+
+.. code-block:: python
+
+   @classmethod
+   def determine_variants(cls, exes, version_str):
+       """Return either a variant string, a tuple of a variant string
+       and a dictionary of extra attributes that will be recorded in
+       packages.yaml or a list of those items.
+       """
+
+This method takes as input a list of executables that live in the same prefix and
+share the same version string, and returns either:
+
+1. A variant string
+2. A tuple of a variant string and a dictionary of extra attributes
+3. A list of items matching either 1. or 2. (if multiple specs are detected
+   from the set of executables)
+
+If extra attributes are returned, they will be recorded in ``packages.yaml``
+and be available for later reuse. As an example, the ``gcc`` package will record
+by default the different compilers found and an entry in ``packages.yaml``
+would look like:
+
+.. code-block:: yaml
+
+   packages:
+     gcc:
+       externals:
+       - spec: 'gcc@9.0.1 languages=c,c++,fortran'
+         prefix: /usr
+         extra_attributes:
+           compilers:
+             c: /usr/bin/x86_64-linux-gnu-gcc-9
+             c++: /usr/bin/x86_64-linux-gnu-g++-9
+             fortran: /usr/bin/x86_64-linux-gnu-gfortran-9
+
+This permits, for instance, to keep track of executables that would be named
+differently if built by Spack (e.g. ``x86_64-linux-gnu-gcc-9``
+instead of just ``gcc``).
+
+.. TODO: we need to gather some more experience on overriding 'prefix'
+   and other special keywords in extra attributes, but as soon as we are
+   confident that this is the way to go we should document the process.
+   See https://github.com/spack/spack/pull/16526#issuecomment-653783204
```

Comment:
(question) I wanted to make sure: this PR no longer supports https://github.com/spack/spack/pull/16526#issuecomment-653783204 right? I assumed we would add that in later, but I wanted to check in case you supported it and I missed that.

Reply:
I added back support for it since it wasn't much effort. Users can add "prefix" or "modules" (or in the future any reserved keyword) to the `extra_attributes` in `determine_variants`. These are popped in:

https://github.com/spack/spack/blob/1d356ca89efd8af0c06aab9ea49e4fcd6d1264d2/lib/spack/spack/package.py#L199-L204

and used to construct the spec. In:

https://github.com/spack/spack/blob/1d356ca89efd8af0c06aab9ea49e4fcd6d1264d2/lib/spack/spack/cmd/external.py#L296-L300

`external_path` is used to set the base_dir if present. I'll add a test to check this case.

Reply:
Test added in 956c191